eduhelx_jupyterlab_student/handlers.py

```diff
@@ -281,48 +186,188 @@ async def post(self):
             return
 
         current_assignment_path = student_repo.get_assignment_path(student_repo.current_assignment)
-        stage_files(".", path=student_repo.repo_root)
-        commit_id = commit(
-            submission_summary,
-            submission_description if submission_description else None,
-            path=student_repo.repo_root
-        )
-        push("origin", "master", path=student_repo.repo_root)
+        
+        rollback_id = get_head_commit_id(path=student_repo.repo_root)
+        stage_files(".", path=current_assignment_path)
+        
+        try:
+            commit_id = commit(
+                submission_summary,
+                submission_description if submission_description else None,
+                path=current_assignment_path
+            )
+        except Exception as e:
+            # If the commit fails then unstage the assignment files.
+            reset(".", path=current_assignment_path)
+            self.set_status(500)
+            self.finish(str(e))
+            return
+
         try:
             await self.api.create_submission(
                 student_repo.current_assignment["id"],
                 commit_id
             )
-            self.finish()
-        except requests.exceptions.HTTPError as e:
+        except Exception as e:
+            # If the submission fails create in the API, rollback the local commit to the previous head.
+            reset(rollback_id, path=student_repo.repo_root)
             self.set_status(e.response.status_code)
             self.finish(e.response.text)
+            return
+        
+        # We need to create the submission in the API before we push the changes to the remote,
+        # so that we don't push the stages changes without actually creating a submission for the user
+        # (which would be very misleading)
+        try:
+            push(ORIGIN_REMOTE_NAME, MAIN_BRANCH_NAME, path=current_assignment_path)
+            self.finish()
+        except Exception as e:
+            # Need to rollback the commit if push failed too.
+            reset(rollback_id, path=student_repo.repo_root)
+            self.set_status(500)
+            self.finish(str(e))
 
 
 class SettingsHandler(BaseHandler):
```

Comment:
I think you've added a lot to this that either doesn't belong in a `SettingsHandler` or you should rename it, probably to something like `ConfigurationHandler` or something. What do you think?

Reply:
Not sure what you mean? SettingsHandler does not do very much, it only returns the server version and repository root.

Reply:
You added a lot of methods to this class

Reply:
Oh wait I think those are at the base level, not within the `SettingsHandler`. My bad, it's hard to tell in Github lol